airbyte-integrations/connectors/source-github/source_github/utils.py

```diff
@@ -40,49 +55,94 @@ class MultipleTokenAuthenticatorWithRateLimiter(AbstractHeaderAuthenticator):
+        self._active_token = next(self._tokens_iter)
+
     @property
     def token(self) -> str:
-        while True:
-            token = next(self._tokens_iter)
-            if self._check_token(token):
-                return f"{self._auth_method} {token}"
 
-    def _check_token(self, token: str):
+        token = self.current_active_token
+        return f"{self._auth_method} {token}"
+
+    def _check_token_limits(self, token: str):
         """check that token is not limited"""
-        self._refill()
-        if self._sleep():
-            self._refill()
-        if self._tokens[token].count > 0:
-            self._tokens[token].count -= 1
-            return True
+        headers = {"Accept": "application/vnd.github+json", "X-GitHub-Api-Version": "2022-11-28"}
+        rate_limit_info = (
+            requests.get(
+                "https://api.github.com/rate_limit", headers=headers, auth=TokenAuthenticator(token, auth_method=self._auth_method)
+            )
+            .json()
+            .get("resources")
+        )
+        token_info = self._tokens[token]
+        token_info.update_at = pendulum.now()
+        remaining_info_core = rate_limit_info.get("core")
+        token_info.count_rest, token_info.reset_at_rest = remaining_info_core.get("remaining"), pendulum.from_timestamp(
+            remaining_info_core.get("reset")
+        )
+
+        remaining_info_graphql = rate_limit_info.get("graphql")
+        token_info.count_graphql, token_info.reset_at_graphql = remaining_info_graphql.get("remaining"), pendulum.from_timestamp(
+            remaining_info_graphql.get("reset")
+        )
 
-    def _refill(self):
-        """refill all needed tokens"""
-        now = time.time()
-        for token, ns in self._tokens.items():
-            if now - ns.update_at >= self.DURATION:
-                ns.update_at = now
-                ns.count = self._requests_per_hour
-
-    def _sleep(self):
-        """sleep only if all tokens is exhausted"""
-        now = time.time()
-        if sum([ns.count for ns in self._tokens.values()]) == 0:
-            sleep_time = self.DURATION - (now - min([ns.update_at for ns in self._tokens.values()]))
-            logging.warning("Sleeping for %.1f seconds to enforce the limit of %d requests per hour.", sleep_time, self._requests_per_hour)
-            time.sleep(sleep_time)
+    def check_all_tokens(self):
+        for token in self._tokens:
+            self._check_token_limits(token)
+
+    def process_token(self, current_token, count_attr, reset_attr):
+        if getattr(current_token, count_attr) > 0:
+            setattr(current_token, count_attr, getattr(current_token, count_attr) - 1)
             return True
+        elif all(getattr(x, count_attr) == 0 for x in self._tokens.values()):
+            min_time_to_wait = min((getattr(x, reset_attr) - pendulum.now()).in_seconds() for x in self._tokens.values())
+            if min_time_to_wait < 600:  # see HttpStream.max_time
```

Comment:
can we store the hardcoded values somewhere as a reference? like what about this `600`?

Reply:
refactored, need this to be merged first: https://github.com/airbytehq/airbyte/pull/34550

Reply:
refactored